src/rust/ide/src/controller/ide.rs

```diff
@@ -0,0 +1,147 @@
+    /// be removed from any "background processes" list.
+    ProcessFinished { handle:ProcessHandle },
+}
+
+/// A publisher for status notification events.
+#[derive(Clone,CloneRef,Debug,Default)]
+pub struct StatusNotificationPublisher {
+    publisher           : notification::Publisher<StatusNotification>,
+    next_process_handle : Rc<Cell<usize>>,
+}
+
+impl StatusNotificationPublisher {
+    /// Constructor.
+    pub fn new() -> Self { default() }
+
+    /// Publish a new status event (see [`StatusNotification::Event`])
+    pub fn publish_event(&self, label:impl Into<String>) {
+        let label        = label.into();
+        let notification = StatusNotification::Event {label};
+        executor::global::spawn(self.publisher.publish(notification));
+    }
+
+    /// Publish a notification about new process (see [`StatusNotification::ProcessStarted`]).
+    ///
+    /// Returns the handle to be used when notifying about process finishing.
+    pub fn publish_process(&self, label:impl Into<String>) -> ProcessHandle {
+        let label  = label.into();
+        let handle = self.next_process_handle.get();
+        self.next_process_handle.set(handle + 1);
+        let notification = StatusNotification::ProcessStarted {label,handle};
+        executor::global::spawn(self.publisher.publish(notification));
+        handle
+    }
+
+    /// Publish a notfication that process has finished (see [`StatusNotification::ProcessFinished`])
+    pub fn published_process_finished(&self, handle:ProcessHandle) {
+        let notification = StatusNotification::ProcessFinished {handle};
+        executor::global::spawn(self.publisher.publish(notification));
+    }
+
+    /// The asynchronous stream of published notifications.
+    pub fn subscribe(&self) -> impl Stream<Item=StatusNotification> {
+        self.publisher.subscribe()
+    }
+}
+
+
+
+// ====================
+// === Notification ===
+// ====================
+
+/// Notification of IDE Controller.
+///
+/// In contrast to [`StatusNotification`], which is a notification from any application part to
+/// be delivered to User (displayed on some event log or status bar), this is a notification to be
+/// used internally in code.
+#[derive(Copy,Clone,Debug)]
+pub enum Notification {
+    /// User created a new project. The new project is opened in IDE.
+    NewProjectCreated
+}
+
+
+
+// ===========
+// === API ===
+// ===========
+
+/// The API of all project management operations.
+///
+/// It is a separate trait, because those methods  are not supported in some environments (see also
+/// [`API::manage_projects`]).
+pub trait ManagingProjectAPI {
+
+    /// Create a new unnamed project and open it in the IDE.
+    fn create_new_project(&self) -> BoxFuture<FallibleResult>;
+}
+
+/// The API of IDE Controller.
+#[automock]
+pub trait API:Debug {
+    /// The model of currently opened project.
+    fn current_project(&self) -> model::Project;
```

Comment:
So we still assume that IDE has exactly one project opened. I wonder if this should be documented in the trait.